src/db/delete_crate.rs

```diff
@@ -14,22 +14,89 @@ enum CrateDeletionError {
 }
 
 pub fn delete_crate(conn: &Connection, name: &str) -> Result<(), Error> {
-    let crate_id_res = conn.query("SELECT id FROM crates WHERE name = $1", &[&name])?;
-    let crate_id = if crate_id_res.is_empty() {
-        return Err(CrateDeletionError::MissingCrate(name.into()).into());
-    } else {
-        crate_id_res.get(0).get("id")
-    };
+    let crate_id = get_id(conn, name)?;
+    delete_crate_from_database(conn, name, crate_id)?;
+    if let Some(s3) = s3_client() {
+        for prefix in STORAGE_PATHS_TO_DELETE {
+            delete_prefix_from_s3(&s3, &format!("{}/{}/", prefix, name))?;
+        }
+    }
+    Ok(())
+}
+
+pub fn delete_version(conn: &Connection, name: &str, version: &str) -> Result<(), Error> {
+    delete_version_from_database(conn, name, version)?;
 
-    delete_from_database(conn, name, crate_id)?;
     if let Some(s3) = s3_client() {
-        delete_from_s3(&s3, name)?;
+        for prefix in STORAGE_PATHS_TO_DELETE {
+            delete_prefix_from_s3(&s3, &format!("{}/{}/", prefix, name))?;
+        }
+    }
+
+    Ok(())
+}
+
+fn get_id(conn: &Connection, name: &str) -> Result<i32, Error> {
+    let crate_id_res = conn.query("SELECT id FROM crates WHERE name = $1", &[&name])?;
+    if let Some(row) = crate_id_res.into_iter().next() {
+        Ok(row.get("id"))
+    } else {
+        Err(CrateDeletionError::MissingCrate(name.into()).into())
     }
+}
 
+// metaprogramming!
+// NOTE: it is _crucial_ that table_name and column_name be hard-coded, i.e. NOT user input
```

Comment:
This makes me a bit worried.

The `delete_metadata` function is only used in `delete_version_from_database`: we could inline it, which would remove the risk of other code miscalling it.

Reply:
Good idea, done.